network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/NodeBreakerViewImpl.java

```diff
@@ -125,44 +126,110 @@ public void traverse(int[] nodes, VoltageLevel.NodeBreakerView.TopologyTraverser
+        return traverseFromNode(graph, node, traversalType, traverser, done);
     }
 
-    private boolean traverseFromNode(Graph<Integer, Edge> graph, int node, VoltageLevel.NodeBreakerView.TopologyTraverser traverser,
+    private boolean traverseFromNode(Graph<Integer, Edge> graph, int node, TraversalType traversalType, VoltageLevel.NodeBreakerView.TopologyTraverser traverser,
                                      Set<Integer> done) {
-        if (done.contains(node)) {
-            return true;
-        }
-        done.add(node);
-
-        for (Edge edge : graph.edgesOf(node)) {
-            NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
-            int nextNode = biConnectable.getNode1() == node ? biConnectable.getNode2() : biConnectable.getNode1();
-            TraverseResult result;
-            if (done.contains(nextNode)) {
-                continue;
-            }
-            if (biConnectable instanceof SwitchAttributes) {
-                result = traverseSwitch(traverser, biConnectable, node, nextNode);
-            } else if (biConnectable instanceof InternalConnectionAttributes) {
-                result = traverser.traverse(node, null, nextNode);
-            } else {
-                throw new AssertionError();
+        if (traversalType == TraversalType.DEPTH_FIRST) {   // traversal by depth first
+            if (done.contains(node)) {
+                return true;
             }
-            if (result == TraverseResult.CONTINUE) {
-                if (!traverseFromNode(graph, nextNode, traverser, done)) {
+            done.add(node);
+
+            for (Edge edge : graph.edgesOf(node)) {
+                NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
+                int nextNode = biConnectable.getNode1() == node ? biConnectable.getNode2() : biConnectable.getNode1();
+                TraverseResult result;
+                if (done.contains(nextNode)) {
+                    continue;
+                }
+                if (biConnectable instanceof SwitchAttributes) {
+                    result = traverseSwitch(traverser, biConnectable, node, nextNode);
+                } else if (biConnectable instanceof InternalConnectionAttributes) {
+                    result = traverser.traverse(node, null, nextNode);
+                } else {
+                    throw new AssertionError();
+                }
+                if (result == TraverseResult.CONTINUE) {
+                    if (!traverseFromNode(graph, nextNode, traversalType, traverser, done)) {
+                        return false;
+                    }
+                } else if (result == TraverseResult.TERMINATE_TRAVERSER) {
                     return false;
                 }
-            } else if (result == TraverseResult.TERMINATE_TRAVERSER) {
-                return false;
             }
+        } else {   // traversal by breadth first
+            boolean keepGoing = true;
```

Comment:
we should be able to get rid of the boolean keepGoing, see proposed changes. Instead of encounteredEdges, we could keep track of visitedNodes similarly to the "done" set in the DFS implementation. See suggestion